src/org/rascalmpl/util/maven/Artifact.java

```diff
@@ -130,70 +133,159 @@ private static boolean isVersionRange(String version) {
                     String effectiveVersion = versions.first();
-                    messages = messages.append(Messages.warning("Multiple version ranges found for " + withoutVersion + " are used. " 
-                        + "It is better to lock the desired version in your own pom to a specifick version, for example <version>["+ effectiveVersion + "]</version>", d.getPomLocation()));
+                    messages = messages.append(Messages.warning("Multiple version ranges found for " + versionLess
+                        + " are used. "
+                        + "It is better to lock the desired version in your own pom to a specifick version, for example <version>["
+                        + effectiveVersion + "]</version>", d.getPomLocation()));
                 }
             }
 
-            if (alreadyIncluded.contains(withoutVersion) || !d.shouldInclude(forScope)) {
+            String resolvedVersion = resolvedVersions.get(versionLess);
+            if (resolvedVersion != null) {
+                alreadyResolved = true;
+                coordinate = new ArtifactCoordinate(coordinate.getGroupId(), coordinate.getArtifactId(), resolvedVersion, coordinate.getClassifier());
+            } else {
+                alreadyResolved = false;
+                resolvedVersions.put(versionLess, version);
+            }
+
+            var key = new ResolveKey(coordinate, excludes);
+            if (alreadySolved.contains(key) || !d.shouldInclude(forScope)) {
                 continue;
             }
+            alreadySolved.add(key);
+
             if (d.getScope() == Scope.PROVIDED) {
                 // current maven behavior seems to be:
                 // - do not download provided dependencies
                 // - if a provided dependency is present in the maven repository it's considered "provided"
                 var pomDep = ourResolver.calculatePomPath(d.getCoordinate());
                 if (!Files.notExists(pomDep)) {
                     // ok, doesn't exist yet. so don't download it to calculate dependencies
-                    // and don't add it to the list since somewhere else somebody might depend on it 
+                    // and don't add it to the list since somewhere else somebody might depend on it
                     continue;
                 }
             }
             if (d.getScope() == Scope.SYSTEM) {
-                result.add(createSystemArtifact(d));                
+                if (!alreadyResolved) {
+                    result.add(createSystemArtifact(d));
+                }
                 continue;
             }
 
-            var art = parser.parseArtifact(coordinate, d.getExclusions(), ourResolver);
+            Set<WithoutVersion> newExcludes = excludes;
+            if (!d.getExclusions().isEmpty()) {
+                newExcludes = new HashSet<>(excludes);
+                newExcludes.addAll(d.getExclusions());
+            }
```

Comment:
Maybe we should wrap the excludes in a immutable set, (`Collections.unmodifiableSet`) to avoid anyone changing the underlying set? as that is what it looks like we're protecting against here?

Reply:
That is not what I am protecting against here, I just need a new set that is the union of the parent excludes and our own excludes. The resulting set should indeed not be modified so protecting against modification is good defensive programming!